Python: Check for insecure versions of SSL and TLS.

taus-semmle opened the pull request:
~Still requires docs and change note. These will be added soon.~

markshannon commented on python/ql/src/Security/CWE-327/InsecureProtocol.ql:
```diff
@@ -0,0 +1,60 @@
+}
+
+private ModuleObject the_pyOpenSSL_module() {
+    result = any(ModuleObject m | m.getName() = "pyOpenSSL").getAttribute("SSL")
```
`result.getName() = "pyOpenSSL.SSL"` would be simpler

markshannon commented on python/ql/src/Security/CWE-327/InsecureProtocol.ql:
```diff
@@ -0,0 +1,60 @@
+    unsafe_ssl_wrap_method_call(call) and method_name = "ssl.wrap_socket"
+or
+    unsafe_pyOpenSSL_Context_call(call) and method_name = "pyOpenSSL.SSL.Context"
+select call, "Insecure SSL/TLS protocol version specified in call to " + method_name + "."
```
It would be nicer to tell the user what version is specified.

markshannon commented on python/ql/test/query-tests/Security/CWE-327/InsecureProtocol.py:
```diff
@@ -0,0 +1,35 @@
+import ssl
+from pyOpenSSL import SSL
```
Can you add tests for `from ssl import PROTOCOL_SSLv2` and similar.

taus-semmle replied:
Done (at bottom of file).

markshannon commented on python/ql/src/Security/CWE-327/InsecureProtocol.ql:
```diff
@@ -0,0 +1,60 @@
+predicate unsafe_ssl_wrap_method_call(CallNode call) {
+    call = ssl_wrap_socket().getACall() and
+    exists(ControlFlowNode arg | arg = call.getArgByName("ssl_version") |
+        arg.(AttrNode).getObject(insecure_version_name()).refersTo(the_ssl_module())
```
`arg.refersTo(the_ssl_module().getAttribute(insecure_version_name())` would be better.

taus-semmle replied:
For this and the above comment, I'm not sure that'll work unless I also stub the `ssl` module. It seems many of the old insecure protocols have been deprecated (and for good reason), hence the use of a more "syntactic" check. I can add the `refersTo` in addition to the syntactic check, however.

markshannon replied:
Are the "METHOD" constants the correct names. For Python 2.7.12
```
>>> [ n for n in dir(ssl) if 'METHOD' in n ]
[]
```

taus-semmle replied:
`METHOD` is for `pyOpenSSL`.
```ql
>>> import ssl
>>> [n for n in dir(ssl) if 'PROTOCOL' in n]
['ALERT_DESCRIPTION_PROTOCOL_VERSION', 'PROTOCOL_SSLv23', 'PROTOCOL_TLS', 'PROTOCOL_TLSv1', 'PROTOCOL_TLSv1_1', 'PROTOCOL_TLSv1_2', '_PROTOCOL_NAMES']
```

markshannon replied:
In which case the comments in `insecure_version_name()` are the wrong way around.

pavgust commented:
This pull request **introduces 2 alerts** when merging 4e2e0e2e6e51dfb17e6dcc1e4e8943042f610f79 into 83d4fb67112d97a9ca43280a439ab6f3d666bf91 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-8f5df2b6e28db82d81b4b7541accbfbb1e8c3431)

**new alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong name for an argument in a class instantiation

---

*Comment posted by [LGTM.com](https://lgtm.com)*

taus-semmle commented:
@felicity-semmle for the new docs.

pavgust commented:
This pull request **introduces 2 alerts** when merging 4e2e0e2e6e51dfb17e6dcc1e4e8943042f610f79 into 453529e3bfaa3c75a5baff5b77344efcf875697b - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-0548aaba271eaf2bdf50155f5bfff04b09f54af2)

**new alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong name for an argument in a class instantiation

---

*Comment posted by [LGTM.com](https://lgtm.com)*

markshannon commented on python/ql/src/Security/CWE-327/InsecureProtocol.ql:
```diff
@@ -0,0 +1,92 @@
+    result = "TLSv1_METHOD" or
+    // For the `ssl` module
+    result = "PROTOCOL_SSLv2" or
+    result = "PROTOCOL_SSLv3" or
```
Should this include `PROTOCOL_SSLv23` given that you are including `SSLv23_METHOD` for `pyOpenSSL.SSL`

adityasharad commented:
What is the status of this PR? I have added the 1.19 milestone, assuming we want this in the release.

taus-semmle commented:
Ah, I was just going to address this. After discussion with Mark, we've decided to retarget this for the next release instead. I will retarget against `next` once I have addressed the comments on documentation (and added a change note).

taus-semmle commented:
My latest commit should address the documentation review comments. Once the changes pass review, I'll retarget against `next` in order to avoid confusion about whether this will go in `rc/1.19` or not. I'll also address Mark's comments and add a change note for 1.20.

felicity-semmle commented:
Apologies for the delay in coming back to this PR. Thanks for the docs changes. Text all LGTM now.

taus-semmle commented on python/ql/src/Security/CWE-327/InsecureProtocol.qhelp:
```diff
@@ -0,0 +1,49 @@
+       <li>Wikipedia: <a href="https://en.wikipedia.org/wiki/Transport_Layer_Security"> Transport Layer Security</a>.</li>
+       <li>Python 3 documentation: <a href="https://docs.python.org/3/library/ssl.html#ssl.SSLContext"> class ssl.SSLContext</a>.</li>
+       <li>Python 3 documentation: <a href="https://docs.python.org/3/library/ssl.html#ssl.wrap_socket"> ssl.wrap_socket</a>.</li>
+       <li>pyOpenSSL documentation: <a href="https://pyopenssl.org/en/stable/api/ssl.html"> An interface to the SSL-specific parts of OpenSSL</a>.</li>
```
I forgot to answer this question. The reference you point out is indeed _not_ relevant for the other query.

taus-semmle commented:
I have retargeted against `master` and added the items I mentioned in my previous comment (missing protocols and a change note). These changes are in the latest two commits -- all other commits are unchanged.

pavgust commented:
This pull request **introduces 2 alerts** when merging f9c6b0eedc6dbe602c438ebdaadcb9f53bb74770 into 97fc4b0c8e254f9661eeee6ae0078e96ba818af6 - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-517b108296e2d4ce190e2cbb66058773b6414ab7)

**new alerts:**

* 1 for Redundant ToString\(\) call
* 1 for Reference equality test on System\.Object

---

*Comment posted by [LGTM.com](https://lgtm.com)*

felicity-semmle commented on change-notes/1.20/analysis-python.md:
```diff
@@ -0,0 +1,29 @@
+ | **Query**                   | **Tags**  | **Purpose**                                                        |
+|-----------------------------|-----------|--------------------------------------------------------------------|
+| Default version of SSL/TLS may be insecure (`py/insecure-default-protocol`) | security, external/cwe/cwe-327 | Results are shown on LGTM by default. |
+| Use of insecure SSL/TLS version (`py/insecure-protocol`) | security, external/cwe/cwe-327 | Results are shown on LGTM by default. |
```
Please add a very brief description of the purpose of each of these queries to the **Purpose** column.

taus-semmle replied:
Ah, I completely forgot. Will do.

felicity-semmle commented on change-notes/1.20/analysis-python.md:
```diff
@@ -0,0 +1,29 @@
+ ## Changes to existing queries
+
+ All taint-tracking queries now support visualization of paths in QL for Eclipse.
+Most security alerts are now visible on LGTM by default.
```
I think that the two lines above may be leftover from the 1.19 release notes. Please can you check?

taus-semmle replied:
You're absolutely right. Thanks for pointing this out.

felicity-semmle commented:
Thanks for the fixes to the change notes. Text in this PR LGTM now.

taus-semmle commented:
Can we merge this, @mark? I would like to get the 1.20 change note in (to avoid having too many conflicts with other PRs in the pipeline).
